nautobot/extras/management/__init__.py

```diff
@@ -77,14 +81,213 @@
     "Planned": "Unit has been planned",
+    "Primary": "Unit is primary",
     "Provisioning": "Circuit is being provisioned",
     "Reserved": "Unit is reserved",
     "Retired": "Location has been retired",
+    "Secondary": "Unit is secondary",
     "SLAAC": "Dynamically assigned IPv6 address",
     "Staged": "Unit has been staged",
     "Staging": "Location is in the process of being staged",
 }
 
+# List of 2-tuples of (model_path, choiceset)
+# Add new mappings here as other models are supported.
+ROLE_CHOICESET_MAP = {
+    "extras.ContactAssociation": extras_choices.ContactAssociationRoleChoices,
+}
+
+# Map of role name -> default hex_color used when importing color choices in `export_roles_from_choiceset()`.
+# Only a small subset of colors are used by default as these originally were derived from Bootstrap CSS classes.
+ROLE_COLOR_MAP = {
+    "Administrative": ColorChoices.COLOR_BLUE,
+    "Billing": ColorChoices.COLOR_GREEN,
+    "Support": ColorChoices.COLOR_YELLOW,
+    "On Site": ColorChoices.COLOR_BLACK,
+}
+
+# Map of role name -> description used when importing role choices in `export_roles_from_choiceset()`.
+ROLE_DESCRIPTION_MAP = {
+    "Administrative": "Unit plays an administrative role",
+    "Billing": "Unit plays a billing role",
+    "Support": "Unit plays a support role",
+    "On Site": "Unit plays an on site role",
+}
+
+
+#
+# Roles
+#
+
+
+def populate_role_choices(apps=global_apps, schema_editor=None, **kwargs):
+    """
+    Populate `Role` model choices.
+
+    This will run the `create_custom_roles` function during data migrations.
+
+    When it is ran again post-migrate will be a noop.
+    """
+    create_custom_roles(apps=apps, **kwargs)
+
+
+def export_roles_from_choiceset(choiceset, color_map=None, description_map=None):
+    """
+    e.g. `export_roles_from_choiceset(ContactAssociationRoleChoices, content_type)`
+
+    This is called by `extras.management.create_custom_roles` for use in
+    performing data migrations to populate `Role` objects.
+    """
+
+    if color_map is None:
+        color_map = ROLE_COLOR_MAP
+    if description_map is None:
+        description_map = ROLE_DESCRIPTION_MAP
+
+    choices = []
+
+    for _, value in choiceset.CHOICES:
+        choice_kwargs = {
+            "name": value,
+            "description": description_map[value],
+            "color": color_map[value],
+        }
+        choices.append(choice_kwargs)
+
+    return choices
+
+
+def create_custom_roles(
+    verbosity=2,
+    apps=global_apps,
+    models=None,
+    **kwargs,
+):
+    """
+    Create database Role choices from choiceset enums.
+
+    This is called during data migrations for importing `Role` objects from
+    `ChoiceSet` enums in flat files.
+    """
+
+    # Only print a newline if we have verbosity!
+    if verbosity > 0:
+        print("\n", end="")
+
+    if "test" in sys.argv:
+        # Do not print output during unit testing migrations
+        verbosity = 1
```

Comment:
Wonder why this is different from the reverse below?

```suggestion
    if "test" in sys.argv:
        # Do not print output during unit testing migrations
        verbosity = 0
        
        if verbosity > 0:
        print("\n", end="")
```